agent/agent.go

```diff
@@ -1253,6 +1278,133 @@ func (a *agent) startReportingConnectionStats(ctx context.Context) {
 	}
 }
 
+var prioritizedProcs = []string{"coder"}
+
+func (a *agent) manageProcessPriorityLoop(ctx context.Context) {
+	if val := a.envVars[EnvProcMemNice]; val == "" || runtime.GOOS != "linux" {
+		a.logger.Info(ctx, "process priority not enabled, agent will not manage process niceness/oom_score_adj ",
+			slog.F("env_var", EnvProcMemNice),
+			slog.F("value", val),
+			slog.F("goos", runtime.GOOS),
+		)
+		return
+	}
+
+	manage := func() {
+		procs, err := a.manageProcessPriority(ctx)
+		if err != nil {
+			a.logger.Error(ctx, "manage process priority",
+				slog.F("dir", agentproc.DefaultProcDir),
+				slog.Error(err),
+			)
+		}
+		if a.modifiedProcs != nil {
+			a.modifiedProcs <- procs
+		}
+	}
+
+	// Do once before falling into loop.
+	manage()
+
+	if a.processManagementTick == nil {
+		ticker := time.NewTicker(time.Second)
+		defer ticker.Stop()
+		a.processManagementTick = ticker.C
+	}
+
+	for {
+		select {
+		case <-a.processManagementTick:
+			manage()
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (a *agent) manageProcessPriority(ctx context.Context) ([]*agentproc.Process, error) {
+	const (
+		procDir     = agentproc.DefaultProcDir
+		niceness    = 10
+		oomScoreAdj = -500
+	)
+
+	procs, err := agentproc.List(a.filesystem, a.syscaller, agentproc.DefaultProcDir)
+	if err != nil {
+		return nil, xerrors.Errorf("list: %w", err)
+	}
+
+	modProcs := []*agentproc.Process{}
+	for _, proc := range procs {
+		// Trim off the path e.g. "./coder" -> "coder"
+		name := filepath.Base(proc.Name())
+		// If the process is prioritized we should adjust
+		// it's oom_score_adj and avoid lowering its niceness.
+		if slices.Contains(prioritizedProcs, name) {
```

Comment:
We want to specifically prioritize the agent and not other coder processes right? If I'm reading this code correctly it would treat `coder server` and `coder stat` the same as the agent.

Reply:
This is a good catch, I don't see that as being a big deal but we can be more discriminate about which processes we want to prioritize by also parsing command arguments. WDYT?

Reply:
why not just check if its the current process?